Fixed #28236 - Integrate dj-database-url into Django

orf opened the pull request:
Ticket: [#28236](https://code.djangoproject.com/ticket/28236)

This PR integrates dj-database-url into Django, allowing users to configure their caches and databases via URLs:

```python
# settings.py
from django.conf.service_url import configure_db, configure_cache, register_db_backend

register_db_backend('customDB', 'my.custom.db.adapter')
register_cache_backend('customCache', 'my.custom.cache.adapter')

# These values would come from environment variables rater than be hard-coded like this

DATABASES = {
    'default': configure_db('postgres://user:password@localhost/myDb?sslmode=verify-full'),
    'sqlite': configure_db('sqlite://:memory:'),
    'custom': configure_db('customDB://10.10.10.10/something?else=1')
}

CACHES = {
    'default': confgure_cache('memcached://127.0.0.1:11211'),
    'file': configure_cache('file:///tmp/cache'),
    'custom': configure_db('customCache://10.10.10.10/something?else=1')
}

```

It's extensible, allowing third party adapters to plug in and extend the parsing. It does this by using a `config_from_url` static method to the base cache adapters and DB wrappers, which allows subclasses to adapt the result - i.e mysql parses some options from the querystring into specific configuration options.

The test suite is copied/adapted from dj-database-url as are a lot of the quirks parsing, so thanks to @kennethreitz for those.

To-Do:
- [ ] Documentation
- [ ] More full-suite tests, ensuring the resulting config actually *works*
- [ ] Add support for easy overriding config options via kwargs

kennethreitz commented:
✨🍰✨

kennethreitz commented:
Let's get some docs added to this, @orf!

orf commented:
I'm a bit busy with my filtered aggregates stuff right now, but I've been
following the discussions on the mailing list. There are some good ideas
I'd like to try first before I make any docs.

I would also really like to figure out a generic way of handling other
settings, like caches, and I've got a few ideas about handing off specific
 parts of the parsing to related classes.

kennethreitz commented:
👍

kennethreitz commented:
If there's anything I can do to help, let me know!

orf commented:
I've finally got around to pushing what I've got so far - sorry about the delay.

So my idea is to add a `configure_from_url` classmethod to `BaseDatabaseWrapper` (and in future to the base cache class). This gets passed a `scheme` and `url` argument.

I think a class based solution is best here, and keeping it inside the database class itself is a good idea. I'll keep working on it, but please let me know what you think.

wkschwartz commented on tests/utils_tests/test_database_url.py:
```diff
@@ -0,0 +1,257 @@
+from django.utils.url_config import configure_db
+
+
+@unittest.skipUnless(connection.vendor == 'postgres', 'Postgres tests')
```
Since these tests don't actually connect to a database, is it possible to remove the `skipUnless` (here and below)?

orf replied:
The downside of this approach is that the DB module needs to be imported, which will fail unless the DB-specific adapters are installed (`psycopg2` etc).

wkschwartz commented on django/db/backends/__init__.py:
```diff
@@ -0,0 +1,12 @@
+from django.utils.url_config import register_db_backend
+
+register_db_backend('mysql', 'django.db.backends.mysql')
```
Could `register_db_backend` be a decorator?

orf replied:
See above, these need to be available pre-import

auvipy commented:
is this going to  be make in django 2.0?

orf commented:
Unfortunately not. There is a fair bit to do on the PR to get it up to scratch.

wkschwartz commented on django/db/backends/oracle/base.py:
```diff
@@ -185,6 +185,12 @@ def __init__(self, *args, **kwargs):
         use_returning_into = self.settings_dict["OPTIONS"].get('use_returning_into', True)
         self.features.can_return_id_from_insert = use_returning_into
 
+    @staticmethod
```
Should this be `classmethod`?

orf replied:
Yep, good catch 👍

wkschwartz commented on django/utils/url_config.py:
```diff
@@ -0,0 +1,65 @@
+    if not hasattr(handler, 'config_from_url'):
+        raise RuntimeError('{0} has no config_from_url method'.format(backend_path))
+
+    if not inspect.ismethod(handler.config_from_url):
```
Would `isinstance(handler.config_from_url, classmethod)` work? If so it has the advantage of not requiring `inspect`.

wkschwartz commented on django/utils/url_config.py:
```diff
@@ -0,0 +1,65 @@
+    handler = module_loading.import_string(backend_path + '.base.DatabaseWrapper')
+
+    if not hasattr(handler, 'config_from_url'):
+        raise RuntimeError('{0} has no config_from_url method'.format(backend_path))
```
Perhaps `TypeError` would be more appropriate here and in the next place below? Having a method is usually a property of a type.

wkschwartz commented on django/utils/url_config.py:
```diff
@@ -0,0 +1,65 @@
+    if scheme in BACKENDS[backend_type]:
+        raise ImproperlyConfigured('Scheme {0} has already been registered as a {1} handler'.format(scheme,
+                                                                                                    backend_type))
+    BACKENDS[backend_type][scheme] = path
```
If a user who writes his own backend makes a typo for `backend_type`, this function will return without error but the registration will mysteriously not have worked. The two options I see are to make a whitelist of allowed `backend_type`s, or make this function private (by adding an underscore in front of the name) and just expose the convenience functions such as `register_db_backend`.

orf replied:
Good point. I've made them private.

wkschwartz commented on django/utils/url_config.py:
```diff
@@ -0,0 +1,65 @@
+BACKENDS = defaultdict(dict)
+
+
+def add_backend(backend_type, scheme, path):
```
This function's name starts with "add" but the convenience functions' names start with "register".  Would it help to keep these uniform?

orf replied:
👍

wkschwartz commented on django/db/backends/postgresql/base.py:
```diff
@@ -144,6 +145,36 @@ def __init__(self, *args, **kwargs):
 
+    @classmethod
+    def config_from_url(cls, engine, scheme, url):
+        parsed = parse.urlparse(url)
```
In the base class's method, it checks whether `URL` is already parsed. Does this class need to do the same check before calling urlparse?

wkschwartz replied:
See also the SQLite backend.

orf replied:
It's a bit iffy, but I added the check to only the base class so that if the method was over-ridden and the `url` was parsed the result could be passed in without needing to re-call `urlparse`. `config_from_url`, when over-ridden, is always going to parse the URL, but the public API should expect only strings to be passed in.

wkschwartz replied:
What if someone wants to subclass the Postgres backend directly?

orf replied:
Hmm. :<

wkschwartz commented on django/db/backends/postgresql/base.py:
```diff
@@ -144,6 +145,36 @@ def __init__(self, *args, **kwargs):
+    @classmethod
+    def config_from_url(cls, engine, scheme, url):
+        parsed = parse.urlparse(url)
+        host, _ = parsed._hostinfo
```
_ is used elsewhere for gettext. Maybe better to just take the [0] slice instead?

orf replied:
👍

wkschwartz commented on tests/utils_tests/test_database_url.py:
```diff
@@ -0,0 +1,208 @@
+    def test_dsn_parsing(self):
+        url = (
+            'oracle://scott:tiger@/'
+            '(DESCRIPTION=(ADDRESS_LIST=(ADDRESS=(PROTOCOL=TCP)'
```
Can you define dsn once and do url += dsn  to reduce duplication?

orf commented:
Thank you for your review @wkschwartz! All your comments seemed very reasonable so I've made the changes requested. If anyone else has any time to review this it would be awesome. Documentation still needs to be added, but this could be ready for 2.0.

orf commented:
I've added support for configuring caches as well. It's a little bit more tricky than database configurations.

I ended up creating a `parse_url` method to deal with common quirks in the standard library `urlparse`, namely case-insensitive hostnames. Still a few more tests to run, but seems to work quite well.

wkschwartz commented on django/utils/url_config.py:
```diff
@@ -0,0 +1,135 @@
+
+BACKENDS = defaultdict(dict)
+
+BaseParsedURL = namedtuple('ParsedURL', 'scheme username password hostname port path options')
```
Could you inherit from [`urllib.parse.ParseResult`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.ParseResult) to make the class hierarchy easier for people already familiar with the stdlib's data types? Then you could do away with the `namedtuple` import and simply do
```python
from urllib import parse
class ParseResult(parse.ParseResult):
    @property
    def netloc(self):
        # ...what you had written below...
```
This should override `ParseResult.netloc` correctly.

orf replied:
Great idea! 👍

orf replied:
Ahh yes. I remember why I did all this, it's because really the ParseResult is pretty unsuited for this due to a lot of annoying quirks. I think it should be a class entirely distinct from ParseResult.

wkschwartz commented on django/utils/url_config.py:
```diff
@@ -0,0 +1,135 @@
+import inspect
```
Would it make more sense to have `url_config` go in `django/conf/`? If you do that, `url_config` might be a confusing name. Maybe something like `services_url_config`? I'm trying to think of something (like "services") that captures both DBs and caches, and the fact that they might be hosted somewhere else.

orf replied:
Yeah I think you're right, I think it should be under `conf`.  `service_url` sounds like a good module name I think?

wkschwartz commented on django/utils/url_config.py:
```diff
@@ -0,0 +1,135 @@
+
+class ParsedURL(BaseParsedURL):
+    @property
+    def netloc(self):
```
Helpful if this should be a public class: A docstring for either `ParsedURL`, `netloc`, or both, explaining what `netloc` does and how `ParsedURL` differs from `urllib.parse.ParseResult`

orf replied:
It should definitely have a doc-string, because I've completely forgotten how it differs....!

wkschwartz commented on django/utils/url_config.py:
```diff
@@ -0,0 +1,135 @@
+    return BACKENDS[backend_type][scheme]
+
+
+def configure(backend_type, value):
```
Should `configure` be private in the same way/for the same reason that `_register_backend` is?

orf replied:
Yep 👍

wkschwartz commented on django/utils/url_config.py:
```diff
@@ -0,0 +1,135 @@
+
+
+def register_db_backend(scheme, path):
+    if not path.endswith('.base.DatabaseWrapper'):
```
I can't find the documentation for how to make a database or cache backend. I'm guessing this is not the only place that explicitly requires this path, but if I'm wrong, should we document this for anyone looking to create their own backends?

wkschwartz commented:
The other thing I just noticed is that `dj-database-url` allows you to pass in option overrides such as `conn_max_age` to the `config` function. Perhaps simple `**kwargs` forwarding from `configure_db` to `configure` to `Database.config_from_url` would do the trick. Then `config_from_url` would simply append kwargs to the `OPTIONS` value, making the key in the kwargs capitalized. I think this will make `configure_db` a drop-in replacement for `dj-database-url`'s `config`.

Thoughts? Is something similar necessary for the cache configurator (at least for uniformity's sake)? (I don't know me much caches.)

orf commented:
I think that needs some thought @wkschwartz, but I agree it's a good feature to have. Implementing overriding individual attributes would be easy, i.e `password='abc', but how we handle nested options (like under 'OPTIONS' etc) would be tricky.

Also, thanks for the review, I've fixed up the branch, rebased it and got the tests to pass. I've made a lot of changes but I think it's a lot neater now. The biggest one was removing the named tuple and instead returning a simple, mutable dictionary from `parse_url`. I've also moved all the code under `django.conf` as it does seem like the right place to have this.

I'll spend some more time fixing up a few small things, any feedback on the code would be much appreciated.

wkschwartz commented on django/core/cache/backends/memcached.py:
```diff
@@ -174,6 +185,15 @@ def _cache(self):
 
     def close(self, **kwargs):
-        # libmemcached manages its own connections. Don't call disconnect_all()
+        # libmemcached managepip install pipenvs its own connections. Don't call disconnect_all()
```
"managepip"?

orf replied:
oh nuts, I accidentally pasted that there and didn't spot it, I'll remove it.

wkschwartz commented:
Maybe we don't even need to support all possible option overrides, just the most common, such as  `conn_max_age`. For more complicated things, we could document this pattern:

```python
DATABASES = {'default': configure_db(os.environ['DATABASE_URL'])}
DATABASES['default']['OPTIONS'] = {'FOO': 'bar'}
```

LuRsT commented:
@orf Need help here?

orf commented:
Hey @wlonk, feel free to take the ticket and have a go. This is on my to-do list but it’s pretty far down, so please anyone who wants to build on this go for it!

felixxm commented:
Closing due to inactivity.
